Update TransitioningContentControl to expose Old Content

fitdev opened the pull request:
<!--- See CONTRIBUTING.md for general guidelines on contributions -->

## What does the pull request do?
Sometimes the `Content` of a `TransitioningContentControl` needs to be disposed of (it implements `IDisposable`). However it is often difficult to manually track of such an old content in the context of `TransitioningContentControl` since the "old" content needs to be "kept alive" until the transition finishes, which only `TransitioningContentControl` knows about, so in general it would be very difficult to manually manage such "old"/"transitioned-away" `Content`.

Therefore a new event has been added, which exposes the previous value of `Content`, which can the be used by a ViewModel to do whatever it needs to do with the old content - like dispose it for example.


## What is the current behavior?
`TransitioningContentControl` does not expose such old `Content` which is no longer needed in any way.


## What is the updated/expected behavior with this PR?
Previous value of the `Content` property is now available via the `TransitionCompletedEvent`, once the transition has been completed and the old content is no longer used.


## Checklist

- [ ] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes
None

## Obsoletions / Deprecations
None

## Fixed issues
N/A

fitdev committed:
Update TransitioningContentControl to support Auto-Disposal of old content

fitdev committed:
Fix typo

fitdev committed:
Fix

MrJul commented:
In my opinion, this solution is a bit too specific.

While I understand the need, I'd prefer if we have either some `TransitionCompleted` event providing the old and new content, or a read-only `OldContent` property which could be observed.

This would allow users to do whatever they want with the old content once the transition has finished: dispose of it, remove it from cache, etc.

fitdev commented:
@MrJul 
Yes, I agree. It would be nice if it was more general. I am not sure how best to implement it though - still new to Avalonia.
I mean events seem to be not very MVVM-friendly.
On the other hand, having `OldContent` property would be MVVM-friendly, but I am not sure how to make it a bindable property however such that the ViewModel won't be able to set it, i.e. taht it would be a "write-only_-like property on the ViewModel updated by the `TransitioningContentControl` every time it is no longer needed.

fitdev committed:
Add OldContent property

fitdev committed:
Make OldContent property read only

maxkatz6 commented on src/Avalonia.Controls/TransitioningContentControl.cs:
```diff
@@ -162,8 +182,10 @@ private void HideOldPresenter()
         {
-            oldPresenter.Content = null;
-            oldPresenter.IsVisible = false;
+          var oldcontent = oldPresenter.Content;
+          oldPresenter.Content = null;
+          oldPresenter.IsVisible = false;
+          OldContent = oldcontent;
```
Couldn't you just handle control.PropertyChanged event, check for Content property and get OldValue from the arguments?
Otherwise, OldContent property feels weird + will hold object in memory more than it should.

fitdev replied:
I was just following all the places in the original design where the `Content` of the "old" presenter was being set to null, which happens in 2 places, not 1. So from my understanding merely handling `PropertyChanged` would not be enough as that would account for just `UpdateContent` path ignoring the `ArrangeOverride` path that also calls `HideOldPresenter`.

> will hold object in memory more than it should

I don't like it too. But how can this be avoided with direct properties?

fitdev replied:
As far as holding object in memory, getter could be modified as such, with documentation explaining that the value is available momentarily:

```cs
    get {
      var v = _OldContent;
      _OldContent = null;
      return v;
    }
```

fitdev replied:
And regarding

> Couldn't you just handle control.PropertyChanged event, check for Content property and get OldValue from the arguments?

If you mean something like this:

```cs
    if (change.Property == ContentProperty) {
      UpdateContent(true);
      //report change.OldValue somehow
      return;
    }
```

Then it does not work because of how `TransitioningContentControl` operates. I tried that approach and started getting weird exceptions related to the fact that I disposed of OldContent in my ViewModel prematurely. So, it looks like the OldContent needs to be set really in the 2 places where originally OldPresenter's content is being set to null - and only there.

maxkatz6 replied:
Have you tried handling OnDetachedFromVisualTree method overload in your content? That's a typical solution for control lifetime - initialize in attached-to-visualtree and dispose in detached-to-visualtree.

fitdev replied:
The problem is that often the "Content" is not an Avalonia object at all, let alone a control. It is an arbitrary DotNet class where there is a `TransitioningContentControl.ContentTemplate` that describes how to actually present such a Content. Hence the need to expose OldContent in a data-bindable-friendly fashion on the `TransitioningContentControl`

fitdev replied:
@maxkatz6 

1. I updated the PR to also account for the missed case when the Old Content goes "out of scope" so to speak and is no longer needed when it happens _during the transition_. It is important to account of all such cases so that Old Content changes can be properly observed by the VM and dealt with as needed.
2. Regarding the backing field for the OldContent property: I do not like it too, but how else can this be addressed? Perhaps the best way would be to instead of doing:

```cs
var oldcontent = oldPresenter.Content;
oldPresenter.Content = null;
OldContent = oldcontent;
```

do this:

```cs
var oldcontent = oldPresenter.Content;
oldPresenter.Content = null;
OldContent = oldcontent;
OldContent = null; //We have signaled to the VM that the old content was no longer needed, so we can now set it to null alltogether - it should have already been taken care of by whoever handled the above line
```

fitdev replied:
Not sure why seemingly unrelated tests failed. I am new to this, so if anyone can clarify, that would be nice.

fitdev committed:
Merge branch 'master' into master

fitdev committed:
Account for all cases of OldContent

Account for the case when OldContent goes out of scope if we are in the middle of transition.

fitdev committed:
Update TransitioningContentControlTests.cs

To make it compatible with the new behavior of nullifying OldContent as soon as we don;t need it.

fitdev committed:
Update TransitioningContentControl.cs

Refactor stuff and nullify OldContent to preserve memory once we have notified of it.

fitdev committed:
Typo fix

Meloman19 commented:
> @MrJul Yes, I agree. It would be nice if it was more general. I am not sure how best to implement it though - still new to Avalonia. I mean events seem to be not very MVVM-friendly. On the other hand, having `OldContent` property would be MVVM-friendly, but I am not sure how to make it a bindable property however such that the ViewModel won't be able to set it, i.e. taht it would be a "write-only_-like property on the ViewModel updated by the `TransitioningContentControl` every time it is no longer needed.

Replace with read only boolean property, which will be an indicator that a transition is run. "bool IsTransitionRunning" for example. Before transition will set true and on complete set false.
This property can be used in MVVM to track the end of the transition and disposed the old model. It can also be useful in the interface, for example, to hide or disable some control during the transition.

fitdev commented:
@Meloman19 This complicates things for the end user as it requires keeping a reference of the old content somewhere. Not to mention, it would be misleading due to the case when multiple transitions are happening one after another without releasing old presenter. So current approach with exposing OldContent is better, and memory is not an issue, since the backing field is set to null once VMs have been notified.

fitdev committed:
Merge branch 'master' into master

fitdev committed:
Merge branch 'master' into master

fitdev committed:
Alternative impl to try fix failing test

fitdev committed:
Merge branch 'master' into master

fitdev committed:
Fix tests after update

fitdev committed:
Fix Test 2

fitdev committed:
Merge branch 'master' into master

fitdev committed:
Merge branch 'master' into master

MrJul commented on src/Avalonia.Controls/TransitioningContentControl.cs:
```diff
@@ -36,6 +36,14 @@ public class TransitioningContentControl : ContentControl
             nameof(IsTransitionReversed),
             defaultValue: false);
 
+  /// <summary>
```
Indent size is wrong it should be 4 spaces, as specified in the .editorconfig (I'm adding only one comment about this, but all added code is affected).

MrJul commented on src/Avalonia.Controls/TransitioningContentControl.cs:
```diff
@@ -55,6 +63,15 @@ public bool IsTransitionReversed
+    get => _OldContent;
+    private set => SetAndRaise(OldContentProperty, ref _OldContent, value);
+  }
+  object? _OldContent;
```
Naming: should be `_oldContent`

fitdev replied:
Thank you for the remarks about style. Will fix it. Not sure what was wrong with indent, I guess GitHub does not handle it too well when editing.

fitdev replied:
And about i being an event. Perhaps - I am still relatively new to Avalonia and not too familiar on how to implement events. So if you can suggest a concrete edit, it would be nice.

MrJul commented on src/Avalonia.Controls/TransitioningContentControl.cs:
```diff
@@ -162,10 +178,20 @@ private void HideOldPresenter()
+      var oldcontent = oldPresenter.Content;
+      oldPresenter.Content = null;
+      OldContent = oldcontent;
+      _OldContent = null; //We have signaled to any bound ViewModel that the old content was no longer needed when we set it in the above line, so we can now set it to null altogether - it should have already been taken care of by whatever logic the ViewModel has for observing the OldContent property.
```
This is wrong: `OldContent` has been changed to `null` but nobody has been notified about it.

fitdev replied:
This is done on purpose, because there is no point of notifying about the fact that the OldContent is no longer relevant. That is done when we set OldContent in the first place. I.e. When we set OldContent, observers get notified and dispose of it (or do whatever they need). At that point the OldContent is effectively gone from the TransitioningControl.

MrJul replied:
Note: I know this has changed since your comment, but there was a point: implementing an Avalonia property implies that changes must be notified. For correctness, you can't go changing a property from A to B to C and notify only A to B and not B to C, even if you deem the C value irrelevant. For example any binding acting on it would still reference the old property value.

MrJul commented on src/Avalonia.Controls/TransitioningContentControl.cs:
```diff
@@ -136,9 +153,8 @@ private void UpdateContent(bool withTransition)
-            _lastPresenter.Content == Content)
-            _lastPresenter.Content = null;
+            _lastPresenter != currentPresenter)
+            NotifyOfOldContent(_lastPresenter);
```
This will completely remove the transition since the content control has just been setup - this should only be done when the transition is complete.

fitdev replied:
Right, this was my misunderstanding. But I still don't understand why `_lastPresenter.Content == Content` was needed? When is it ever the case?

fitdev replied:
Actually my original design was this in the `ArrangeOverride`:

```cs
        transition.Start(from, to, !IsTransitionReversed, cancel.Token).ContinueWith(x => {
          if (!cancel.IsCancellationRequested) {
            HideOldPresenter();
          }
          else if (from is ContentPresenter oldPresenter) {
            //Even though we are not hiding old presenter yet, we still need to notify that its Content (OldContent) is no longer needed and can be disposed of if need be:
            //NotifyOfOldContent(oldPresenter);
          }
        }, TaskScheduler.FromCurrentSynchronizationContext());
```

MrJul commented on tests/Avalonia.Controls.UnitTests/TransitioningContentControlTests.cs:
```diff
@@ -167,7 +167,7 @@ public void New_Transition_Should_Be_Started_If_Content_Changes_While_Running()
                 Assert.Same(presenter2, fromPresenter);
                 Assert.Same(target.Presenter, toPresenter);
-                Assert.Equal("bar", fromPresenter.Content);
+                Assert.Equal(null, fromPresenter.Content); //We explicitly nullify old Content as soon as we are done with it so that it can be safely disposed by any ViewModel observing OldContent property
```
As mentioned on previous comments, this doesn't seem right: the transition has just started and hasn't been complete, the content shouldn't be null already. These tests shouldn't be changed.

fitdev replied:
I believe this is correct because we are tackling the case of Content being changed _while the transition is running_, and so we therefore have 3 "Contents" the "From" and "To" Contents in the running transition, and the newest Content which has just been set.

So the logic is, we should ignore the current transition (it should be cancelled with cancellation token in such a case), and then proceed with the newest Content's logic (i.e. we set the "From"'s content of the running transition to null (the oldest of them all)).

fitdev replied:
The main issue I was trying to tackle is how to properly notify of OldContent which arises when new content is being set _while the transition is still running_ (i.e. 3 "Contents" situation).

MrJul replied:
No, this isn't correct: just try it in the `ControlCatalog` and see the old content missing if you try to change the content while the transition is in progress. e.g transitioning from A to B then B to C (while the first transition A to B hasn't finished) gives a transition from null to C.

I believe you're overthinking this a little :) The current logic before this PR already correctly releases the old content and doesn't have to change. The only thing missing is to notify of this old content whenever that happens, that is when any `ContentPresenter.Content` changes.

MrJul reviewed (changes requested):
Sorry, I didn't have time to get back to this PR since my first comment.

As mentioned in the review comments, the content should be set to `null` only when the transition has finished, like it was before. The only method needing change should be `HideOldPresenter`. The transition is currently broken with this PR.

(I mentioned `OldContent` being a possibility in my original post but honestly I find an event easier here. It's easy to handle in a more MVVM way with a XAML behavior if needed.)

fitdev committed:
Reexpose OldContent via Event

fitdev committed:
Fix

fitdev committed:
Update TransitioningContentControlTests.cs

fitdev committed:
Update TransitioningContentControlTests.cs

fitdev committed:
Update TransitioningContentControlTests.cs

fitdev committed:
Update TransitioningContentControlTests.cs

fitdev committed:
Merge branch 'master' into master

MrJul commented:
@fitdev While I appreciate the effort, while trying your changes, I've implemented the requested feature in a new branch, with tests (see upcoming PR).

Also, are you editing the files directly on github? For easier and faster iteration, I would really recommend cloning and testing changes locally, then opening a PR.

fitdev commented:
@MrJul 

Thanks for looking into this. I am new to Avalonia and OS contributions so learning as I go, so sorry for various hickups.

Does your PR address the need to notify of Old Content for cases when transition has been canceled  (i.e. when the Content changes too quickly for a transition to complete)? We still want to be notified in such cases too.

MrJul commented:
> @MrJul
> 
> Thanks for looking into this. I am new to Avalonia and OS contributions so learning as I go, so sorry for various hickups.
No worries about that, first contributions aren't easy!

> Does your PR address the need to notify of Old Content for cases when transition has been canceled (i.e. when the Content changes too quickly for a transition to complete)? We still want to be notified in such cases too.

Yes, it handles this case, there's even a unit test for it.
